public/app/plugins/panel/heatmap/utils.ts

```diff
@@ -585,15 +621,24 @@ export function heatmapPathsDense(opts: PathbuilderOpts) {
         let ySize: number;
 
         if (scaleX.distr === 3) {
-          xSize = Math.abs(valToPosX(xs[0] * scaleX.log!, scaleX, xDim, xOff) - valToPosX(xs[0], scaleX, xDim, xOff));
+          // For log scales, calculate cell size from actual adjacent bucket positions
+          const nextXValue = xs[yBinQty] ?? xs[0] * scaleX.log!;
+          xSize = Math.abs(valToPosX(nextXValue, scaleX, xDim, xOff) - valToPosX(xs[0], scaleX, xDim, xOff));
         } else {
           xSize = Math.abs(valToPosX(xBinIncr, scaleX, xDim, xOff) - valToPosX(0, scaleX, xDim, xOff));
         }
 
         if (scaleY.distr === 3) {
-          ySize =
-            Math.abs(valToPosY(ys[0] * scaleY.log!, scaleY, yDim, yOff) - valToPosY(ys[0], scaleY, yDim, yOff)) /
-            ySizeDivisor;
+          if (ySizeDivisor === 1) {
+            // Use actual data spacing for pre-bucketed data
+            const nextYValue = ys[1] ?? ys[0] * scaleY.log!;
+            ySize = Math.abs(valToPosY(nextYValue, scaleY, yDim, yOff) - valToPosY(ys[0], scaleY, yDim, yOff));
+          } else {
+            // Use full magnitude for calculated heatmaps with splits
+            ySize =
+              Math.abs(valToPosY(ys[0] * scaleY.log!, scaleY, yDim, yOff) - valToPosY(ys[0], scaleY, yDim, yOff)) /
+              ySizeDivisor;
+          }
```

Comment:
Fix log scale cell sizing
Use actual adjacent bucket values (ys[1], xs[yBinQty])
Instead of computed values (ys[0] * log)